src/librustc_mir/monomorphize/partitioning.rs

```diff
@@ -421,12 +423,48 @@ fn merge_codegen_units<'tcx>(initial_partitioning: &mut PreInliningPartitioning<
     // the stable sort below will keep everything nice and deterministic.
     codegen_units.sort_by_key(|cgu| cgu.name().clone());
 
+    // Estimate the size of a codegen unit as (approximately) the number of MIR
+    // statements it corresponds to.
+    fn codegen_unit_size_estimate<'a, 'tcx>(cgu: &CodegenUnit<'tcx>,
+                                            mono_item_sizes: &HashMap<MonoItem, usize>)
+                                            -> usize {
+        cgu.items().keys().map(|mi| mono_item_sizes.get(mi).unwrap()).sum()
+    }
+
+    fn mono_item_size_estimate<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                         item: &MonoItem<'tcx>)
+                                         -> usize {
+        match item {
+            MonoItem::Fn(instance) => {
+                // Estimate the size of a function based on how many statements
+                // it contains.
+                let mir = tcx.instance_mir(instance.def);
+                mir.basic_blocks().iter().map(|bb| bb.statements.len()).sum()
+            },
+            // Conservatively estimate the size of a static declaration
+            // or assembly to be 1.
+            MonoItem::Static(_) | MonoItem::GlobalAsm(_) => 1,
+        }
+    }
+
+    // Since `sort_by_key` currently recomputes the keys for each comparison,
+    // we can save unnecessary recomputations by storing size estimates for
+    // each `MonoItem`. Storing estimates for `CodegenUnit` might be preferable,
```

Comment:
I'm not sure whether it's worth trying to store estimates for `CodegenUnit` instead. I found it somewhat troublesome to use `CodegenUnit` as a key, but perhaps it'd be worth introducing an index for the units to use instead.